server/services/diplomacy.ts

```diff
@@ -143,53 +141,84 @@ export default class DiplomacyService extends EventEmitter {
         return true;
     }
 
-    getFilteredDiplomacy(player: Player, forPlayer: Player): PlayerDiplomacy[] {
-        return player.diplomacy.filter(d => d.playerId.toString() === forPlayer._id.toString());
+    getFilteredDiplomacy(player: Player, forPlayer: Player): PlayerDiplomacy {
+        return {
+            otherPlayers: player.diplomacy.otherPlayers.filter(a => a.toString() === forPlayer._id.toString()),
+            alliancesMadeThisCycle: player.diplomacy.alliancesMadeThisCycle
+        }
     }
 
     async _declareStatus(game: Game, playerId: DBObjectId, playerIdTarget: DBObjectId, state: DiplomaticState, saveToDB: boolean = true) {
-        let player: Player = game.galaxy.players.find(p => p._id.toString() === playerId.toString())!;
+      let player: Player = game.galaxy.players.find(p => p._id.toString() === playerId.toString())!;
 
-        // Get the current status between the two players.
-        let diplo = player.diplomacy.find(d => d.playerId.toString() === playerIdTarget.toString());
+       let allyCount = -1;
+        if (game.settings.alliances.maxAlliances > 0) {
+          allyCount = this.getAlliesOfPlayer(game, player, true).length;
+        }
+
+        let diploStatusBefore = this.getDiplomaticStatusToPlayer(game, playerId, playerIdTarget);
+        let diplo = player.diplomacy.otherPlayers.find(d => d.playerId.toString() === playerIdTarget.toString());
+
+        //only add alliance request if the target player is not already an ally and if the player has capacity for more allies
+        if (state == 'allies' && (allyCount >= game.settings.alliances.maxAlliances || diploStatusBefore.actualStatus == 'allies')) {
+            return diploStatusBefore; // no change
```

Comment:
A `ValidationError` would be more appropriate here.

Reply:
Hrm.. not sure if I can get this to work with a validation error.   

Just throwing a validation error here causes the API request to fail which gets dropped by the client.  By just returning the diplomatic status unchanged, the client is able to check that the status that came back didnt change so it shows the failed to ally toast message.  

Maybe you are better able to revamp this one?